### PR TITLE
optimize webpack runtime size

### PR DESCRIPTION
**Pull request by smelukov**
Optimize webpack runtime size by creating a hasOwnProperty helper and replace function declarations to arrow function if env is supporting it

After compiling all the examples with compression and diff the size of `*/dist/*.js`, I've got:
**Before PR**: 593.6kb
**After PR**: 582.1kb

**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

no need

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

nothing

### CODE REVIEW

File: lib/RuntimeGlobals.js [hunk 1]
```diff
@@ -199,3 +199,9 @@ exports.amdOptions = "__webpack_require__.amdO";
  * the System polyfill object
  */
 exports.system = "__webpack_require__.System";
+
+/**
+ * the shorthand for Object.prototype.hasOwnProperty
+ * using of ot decreases the compiled bundle size
+ */
+exports.hasOwnProperty = "__webpack_require__.hop";
```
Comment (sokra) [inline lib/RuntimeGlobals.js]: Could you use a single letter property name to be even shorter. I think we used `__webpack_require__.o` before for this runtime function.

File: lib/javascript/JavascriptModulesPlugin.js [hunk 2]
```diff
@@ -578,10 +578,17 @@ class JavascriptModulesPlugin {
 					const innerStrict = !allStrict && m.buildInfo.strict;
 					const iife = innerStrict || inlinedModules.size > 1 || chunkModules;
 					if (iife) {
-						source.add("!function() {\n");
-						if (innerStrict) source.add('"use strict";\n');
-						source.add(renderedModule);
-						source.add("\n}();\n");
+						if (runtimeTemplate.supportsArrowFunction()) {
```
Comment (sokra) [inline lib/javascript/JavascriptModulesPlugin.js]: Check for `RuntimeGlobals.thisAsExports` when the module is using `this` (which actually point to the `exports` object in CommonJs)

File: lib/javascript/JavascriptModulesPlugin.js [hunk 3]
```diff
@@ -710,6 +717,24 @@ class JavascriptModulesPlugin {
 			buf.push("");
 		}
 
+		if (runtimeRequirements.has(RuntimeGlobals.hasOwnProperty)) {
```
Comment (sokra) [inline lib/javascript/JavascriptModulesPlugin.js]: This should not be here, but a separate HelperRuntimeModule referenced in `RuntimePlugin.js`

File: lib/javascript/JavascriptModulesPlugin.js [hunk 4]
```diff
@@ -710,6 +717,24 @@ class JavascriptModulesPlugin {
...
+			buf.push("// the shorthand for Object.prototype.hasOwnProperty");
+			if (runtimeTemplate.supportsArrowFunction()) {
```
Comment (sokra) [inline lib/javascript/JavascriptModulesPlugin.js]: You can use `runtimeTemplate.basicFunction` to shorten this

File: lib/web/JsonpChunkLoadingRuntimeModule.js [hunk 5]
```diff
@@ -18,6 +18,7 @@ class JsonpChunkLoadingRuntimeModule extends RuntimeModule {
 		this.jsonpScript = jsonpScript;
 		this.linkPreload = linkPreload;
 		this.linkPrefetch = linkPrefetch;
+		this.runtimeRequirements.add(RuntimeGlobals.hasOwnProperty);
```
Comment (sokra) [inline lib/web/JsonpChunkLoadingRuntimeModule.js]: Don't add it here, but in JsonpTemplatePlugin where this runtime module is added

File: lib/web/JsonpTemplatePlugin.js [hunk 6]
```diff
@@ -178,9 +179,13 @@ class JsonpTemplatePlugin {
 						: "",
 					"// create error before stack unwound to get useful stacktrace later",
 					"var error = new Error();",
-					"onScriptComplete = function (event) {",
+					runtimeTemplate.supportsArrowFunction()
```
Comment (sokra) [inline lib/web/JsonpTemplatePlugin.js]: use runtimeTemplate.basicFunction

File: lib/webworker/ImportScriptsChunkLoadingRuntimeModule.js [hunk 7]
```diff
@@ -12,6 +12,7 @@ class ImportScriptsChunkLoadingRuntimeModule extends RuntimeModule {
 	constructor(runtimeRequirements) {
 		super("importScripts chunk loading", 10);
 		this.runtimeRequirements = runtimeRequirements;
+		this.runtimeRequirements.add(RuntimeGlobals.hasOwnProperty);
```
Comment (sokra) [inline lib/webworker/ImportScriptsChunkLoadingRuntimeModule.js]: move to WebWorkerTemplatePlugin

**Comment by smelukov** [comment]
After compiling all the examples with compression and diff the size of `*/dist/*.js`, I've got:
**Before PR**: 593.6kb
**After PR**: 582.1kb

File: lib/RuntimeTemplate.js [hunk 8]
```diff
@@ -430,29 +437,41 @@ class RuntimeTemplate {
 					weak,
 					runtimeRequirements
 				});
-				getModuleFunction = `function() { ${header}return ${rawModule}; }`;
+				getModuleFunction = this.basicFunction(
+					"",
+					`${header}return ${rawModule}`
```
Comment (sokra) [inline lib/RuntimeTemplate.js]:
```suggestion
					`${header}return ${rawModule};`
```

File: lib/RuntimeTemplate.js [hunk 9]
```diff
@@ -430,29 +437,41 @@ class RuntimeTemplate {
...
 				runtimeRequirements.add(RuntimeGlobals.require);
 				getModuleFunction = `__webpack_require__.bind(null, ${comment}${idExpr})`;
 			}
 		} else if (strict) {
 			runtimeRequirements.add(RuntimeGlobals.createFakeNamespaceObject);
 			if (header) {
-				getModuleFunction = `function() { ${header}return ${RuntimeGlobals.createFakeNamespaceObject}(${moduleIdExpr}, 1); }`;
+				const returnExpression = `${RuntimeGlobals.createFakeNamespaceObject}(${moduleIdExpr}, 1)`;
+				getModuleFunction = header
+					? this.basicFunction("", `${header}return ${returnExpression}`)
```
Comment (sokra) [inline lib/RuntimeTemplate.js]:
```suggestion
					? this.basicFunction("", `${header}return ${returnExpression};`)
```

File: lib/RuntimeTemplate.js [hunk 10]
```diff
@@ -430,29 +437,41 @@ class RuntimeTemplate {
...
 			} else {
 				getModuleFunction = `${RuntimeGlobals.createFakeNamespaceObject}.bind(__webpack_require__, ${comment}${idExpr}, 1)`;
 			}
 		} else if (exportsType === "default") {
 			runtimeRequirements.add(RuntimeGlobals.createFakeNamespaceObject);
 			if (header) {
-				getModuleFunction = `function() { ${header}return ${RuntimeGlobals.createFakeNamespaceObject}(${moduleIdExpr}, 3); }`;
+				const returnExpression = `${RuntimeGlobals.createFakeNamespaceObject}(${moduleIdExpr}, 3)`;
+				getModuleFunction = header
+					? this.basicFunction("", `${header}return ${returnExpression}`)
```
Comment (sokra) [inline lib/RuntimeTemplate.js]:
```suggestion
					? this.basicFunction("", `${header}return ${returnExpression};`)
```

File: lib/RuntimeTemplate.js [hunk 11]
```diff
@@ -430,29 +437,41 @@ class RuntimeTemplate {
...
 			} else {
 				getModuleFunction = `${RuntimeGlobals.createFakeNamespaceObject}.bind(__webpack_require__, ${comment}${idExpr}, 3)`;
 			}
 		} else {
 			runtimeRequirements.add(RuntimeGlobals.createFakeNamespaceObject);
 			if (header) {
-				getModuleFunction = `function() { ${header}return ${RuntimeGlobals.createFakeNamespaceObject}(${moduleIdExpr}, 7); }`;
+				const returnExpression = `${RuntimeGlobals.createFakeNamespaceObject}(${moduleIdExpr}, 7)`;
+				getModuleFunction = header
+					? this.basicFunction("", `${header}return ${returnExpression}`)
```
Comment (sokra) [inline lib/RuntimeTemplate.js]:
```suggestion
					? this.basicFunction("", `${header}return ${returnExpression};`)
```

File: lib/javascript/JavascriptModulesPlugin.js [hunk 12]
```diff
@@ -578,10 +578,20 @@ class JavascriptModulesPlugin {
 					const innerStrict = !allStrict && m.buildInfo.strict;
 					const iife = innerStrict || inlinedModules.size > 1 || chunkModules;
 					if (iife) {
-						source.add("!function() {\n");
-						if (innerStrict) source.add('"use strict";\n');
-						source.add(renderedModule);
-						source.add("\n}();\n");
+						if (
+							!runtimeRequirements.has(RuntimeGlobals.thisAsExports) &&
```
Comment (sokra) [inline lib/javascript/JavascriptModulesPlugin.js]:
```suggestion
```

Actually I told you something wrong. That doesn't make sense here in any way. It's an IIFE and doesn't pass any `this` value anyway.

File: lib/node/ReadFileChunkLoadingRuntimeModule.js [hunk 13]
```diff
@@ -14,6 +14,7 @@ class ReadFileChunkLoadingRuntimeModule extends RuntimeModule {
 	constructor(runtimeRequirements) {
 		super("readFile chunk loading", 10);
 		this.runtimeRequirements = runtimeRequirements;
+		this.runtimeRequirements.add(RuntimeGlobals.hasOwnProperty);
```
Comment (sokra) [inline lib/node/ReadFileChunkLoadingRuntimeModule.js]: move this to NodeTemplatePlugin

File: lib/node/RequireChunkLoadingRuntimeModule.js [hunk 14]
```diff
@@ -14,6 +14,7 @@ class RequireChunkLoadingRuntimeModule extends RuntimeModule {
 	constructor(runtimeRequirements) {
 		super("require chunk loading", 10);
 		this.runtimeRequirements = runtimeRequirements;
+		this.runtimeRequirements.add(RuntimeGlobals.hasOwnProperty);
```
Comment (sokra) [inline lib/node/RequireChunkLoadingRuntimeModule.js]: move this to NodeTemplatePlugin

File: lib/runtime/GlobalRuntimeModule.js [hunk 15]
```diff
@@ -17,9 +17,9 @@ class GlobalRuntimeModule extends RuntimeModule {
 	 * @returns {string} runtime code
 	 */
 	generate() {
+		const { runtimeTemplate } = this.compilation;
 		return Template.asString([
-			`${RuntimeGlobals.global} = (function() {`,
-			Template.indent([
+			`${RuntimeGlobals.global} = ${runtimeTemplate.iife("", [
```
Comment (sokra) [inline lib/runtime/GlobalRuntimeModule.js]: revert this change. It must always be a `function()` because `this` is used in body.

**Comment by sokra** [comment]
Thanks
